### ISSUE TITLE
dart2js: Incorrect inlining around await for elements

### ISSUE THREAD

**Issue by munificent** [dart-lang/sdk#41675]
I see a couple of dart2js failures in [this change](https://dart-review.googlesource.com/c/sdk/+/144305) to a couple of test packages that only happens when those packages are treated as opted in to NNBD, even though the failure itself does not seem directly related to NNBD.

@leafpetersen crafted this repro, which I've winnowed down a bit more:

### opted_out_test.dart:

```dart
// @dart = 2.6
import 'opted_in.dart';

void main() async {
  Stream<int> nullStream = null;
  test(() async => [await for (var i in nullStream) 1]);
}
```

### opted_in.dart:

```dart
import 'dart:async';

void test(Future<void> f()) {
  f().then((_) {
    throw "Should not get here. f() should throw.";
  }).catchError((error, stack) {
    print("Caught $error");
    if (error is! Error) throw "Should not get here, error should be an Error.";
  });
}
```

This test should complete successfully because the error reported on the null stream should be caught by the `catchError()` clause. Instead, it escapes because the throw gets hoisted outside of that context.

I'm guessing this is probably only an issue with `await for` collection elements since those seem to be the only tests broken by the change. It may also only be an issue with code that crosses NNBD boundaries, but that's also just a guess.

**Comment by fishythefish** [comment]
@rakudrama Do you have bandwidth to look at this?

**Comment by rakudrama** [comment]
Fixed via ef1528aa78c646f786717ad6174b521699ece1a2.
The throw did not get hoisted - it was much simpler - there was a bug in the subtype algorithm that caused an 'impossible' Error in the runtime (and outside of the protected region).
